cmd/juju/application/utils/utils.go

```diff
@@ -118,38 +153,87 @@ func filterResources(
 }
 
 // shouldUpgradeResource reports whether we should upload the metadata for the given
-// resource.  This is always true for resources we're adding with the --resource
-// flag. For resources we're not adding with --resource, we only upload metadata
-// for charmstore resources.  Previously uploaded resources stay pinned to the
-// data the user uploaded.
-func shouldUpgradeResource(newCharmURL *charm.URL, res charmresource.Meta, uploads map[string]string, current map[string]resources.Resource) (bool, error) {
-	// Always upload metadata for resources the user is uploading during
-	// upgrade-charm.
-	if _, ok := uploads[res.Name]; ok {
-		logger.Tracef("%q provided to upgrade existing resource", res.Name)
+// resource.
+//
+// Upgrade when:
+//  1. Resources specified as a local file on the cli to be uploaded.
+//  2. Resources specified as resource revision on the cli, which are
+//     not currently in use.
+//  3. Upstream has a new resource available than what is currently used.
+//  4. A new upstream resource is available.
```

Comment:
so long as it has been published right?

Reply:
yes, the repositoryResourceLister won't be given the data if it hasn't been published.